Add authentication to the docker daemon

nalind opened the pull request:
This set of patches implements additions to the docker daemon and client that were outlined in PR #17412.  The daemon expects clients to authenticate if it's started with the -a or --authn flag, and specific requirements are controlled by passing configuration settings to its new --authn-opt option.  The client's behavior when a daemon indicates that it requires clients to authenticate can also be controlled with its --authn-opt option.

nalind committed:
Generate Basic auth data

When a server offers Basic authentication, check for a user name
specified as an authentication option on the command-line, and for both
the user name and password in the environment (preferring the user name
value from the command-line if both are present), and if we don't have
both values, either prompt for or confirm the user name, and prompt for
the password to provide to the daemon.

Cache those answers so that we don't prompt more than once, even if a
given command-line invocation results in several connections and
requests to the daemon.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Call into libsasl to check Basic auth passwords

Use libsasl2 to check passwords if the "libsasl2=true" authn option is
used, offering it as an option for Basic authentication with a Basic
realm name which is either specified using the "realm" authn option, or
the default of "localhost".  The SASL service name used for the password
check is hard-coded to "docker".

Deployments which expect to use saslauthd for checking the passwords
will need to create a file named "docker.conf" under /usr/lib/sasl2 or
/etc/sasl2 containing this line:
  pwcheck_method:saslauthd

NB: When we move to building with a later version of the OS, we should
probably change the cgo directive that adds -lsasl2 to the LDFLAGS to
take advantage of pkg-config data, and update hack/make.sh to match.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Call into gssapi to generate Negotiate auth tokens

Attempt to use the GSSAPI libraries to generate SPNEGO GSSAPI tokens
when the server says that we can use Negotiate.

We set the service name following HTTP conventions, as the "HTTP"
host-based service running on the HTTP server.  We don't convert server
names that look like IP addresses back into names, so those probably
won't succeed unless the KDC is exceptionally clever about handling
service referrals.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Add authentication by asking the kernel

On Linux systems, if the "local-auth=true" authentication option is
used, ask the kernel for the Uid on the other end of a Unix socket
connection, and look up the name of the user.  This provides an implicit
authentication for the common cases.

Note that the lookup is currently only supported for non-static builds,
because we use the standard library's lookup logic, which only returns
an error in static builds.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Vendor github.com/gorilla/{securecookie,sessions}

We'll need these to set cookies for clients, which is how we'll offload
the state of client-to-daemon authentication state to clients, to avoid
having to hold any of that state in the daemon itself.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Set and use cookies for client-to-daemon auth

When we've authenticated a client, set a cookie that the client will be
able to use to avoid having to authenticate any subsequent requests that
it will issue before we exit.  Thanks to Adrian P for help testing this.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Add a client-daemon authentication framework

Add the basic framework for requiring clients to authenticate to the
daemon when issuing requests to it.  This is mainly plumbing and wiring
in the logic in both the client and the server.

The server contains a module-local list of functions which create and
return handlers for authentication schemes.  The implementations
register themselves at initialization time.  On servers, the functions
are used to create a list of Authenticators in each server which has
authentication enabled, each of which gets a crack at every client
request.

The server will call either the GetChallenge methods or CheckResponse
methods of its Authenticators for each request, depending on whether or
not the incoming request includes any Authorization headers.  If
CheckResponse indicates that the client's identity has been verified,
information about the authenticated user, and the name of the
authentication scheme that succeeded, will be added to the request
structure as a shared value so that it can be used later when we call
into authorization plugins.  The middleware for authorization is moved
earlier in order to get the ordering right at run time.

The server's RequireAuthn flag can be found in its configuration
structure, and both client and server get an opportunity to retrieve
option values, which are passed to the --authn-opt flag on the command
line, from a map named AuthnOpts in the server configuration structure.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

Fixed a bug which disabled persistent connections when docker authentication is required

Signed-off-by: lior <lior@twistlock.com>

Commit:
Added kerberos authentication integration tests.

Signed-off-by: Lior Yankovich <lior@twistlock.com>

nalind committed:
Call out to htpasswd to check Basic auth passwords

Run htpasswd to check Basic authenticaiton passwords if the --htpasswd
flag is used to specify a file to use as the password database and we
can find the "htpasswd" command at run-time.  There is no default value,
so the option is disabled by default.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

lioryanko committed:
Improved kerberos tests

Signed-off-by: Lior Yankovich <lior@twistlock.com>

nalind committed:
Fixes for Kerberos tests

Make changes as advised by the validation tests, and revert the bits
that changed the Cmd() and CmdWithArgs() implementations, since we ended
up not needing those to be changed.  Also, add the weak-crypto flag to
our call to 'kdb5_util create' to avoid getting stuck waiting for
entropy while creating a realm database that will only be around while
the tests are running.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Add tests for Authn using the Basic scheme

Add tests to test Basic authentication, using each of libsasl2 (pointed
to saslauthd, using its built-in "kerberos5" mechanism, and reusing most
of the logic for setting up a keytab for Negotiate auth to verify TGTs)
and htpasswd on the server side.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Add tests for kernel-based authentication

Add tests to test that we can accept authentication by asking the kernel
for the UID of the remote end of a Unix socket connection.  Check that
we don't incorrectly think that this can work when a client connects
over TCP.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Add authentication from examining client certs

Add a call to the MapClientCertificateToUser() API provided by one or
more "ClientCertificateMapper" plugins, whose names are passed in as a
list of values of the "certmap" authentication option, and which takes a
binary certificate and returns a user information structure, for mapping
a client's certificate to an user identity as a means of authenticating
clients.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Test authentication from examining client certs

Split out the test for authentication by certificate mapping, using a
dumb test mapper which accepts only one certificate, which it maps to a
statically-defined user.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Add authentication via external plugins

Add a call to the Authentication.GetChallenge() API provided by one or
more "Authentication" plugins, whose names can be passed to the daemon
as a list of values for the "plugins" authentication option, and which
provide headers to send back to a client that sent us a request with no
"Authorization" header.  Call the plugin Authentication.CheckResponse()
methods when the client request includes at least one "Authorization"
header, and if any plugin returns a usable User structure, accept it as
proof of the client's identity.

We can't do the client part of authentication using plugins, because
it's not really useful to require the client to use plugins which have
to be registered in a central location and ensure that a plugin is
running, so the client still has to be taught how to respond to
particular authentication requirements directly.

Some of the server bits we add later in the patch series could be moved
out to plugins.  We're already linking with libraries for the client's
sake, so using a plugin for Negotiate instead of calling the APIs
directly doesn't avoid the linking dependency.  The client socket can't
be passed to a plugin, so looking up the user's identity using the
socket descriptor and some help from the kernel is not something that
can be done via a plugin.

That said, plugins remain a viable method for implementing additional
ways to check passwords for Basic authentiation, for verifying Bearer
tokens, and for supporting other schemes which the connecting client may
support.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Test authentication via external plugins

Implement a dumb "Bearer" plugin to test both the server's
authentication plugin logic and the client's Bearer implementation, and
also take the opportunity to verify that authorizing using the client's
user name is possible.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Documentation for proposed authn features

This patch contains the documentation portions of proposed
authentication changes discussed in PR#13697, previously proposed as a
docs-only change in PR#17412, for PR#18514.  To sum up:
* The daemon requires clients to authenticate if it's started with the
  -a/--authn flag, otherwise it doesn't, preserving the previous
  behavior.
* The specifics of how the daemon authenticates clients, and how the
  client responds to a daemon's requirement for authentication, are
  controlled by option values passed using the --authn-opt flag.  Note
  that clients expect --authn-opt flags to be specified before the
  client command rather than after it.

* The daemon will offer support for HTTP Basic authentication if the
  daemon is started with either the "libsasl2" or "htpasswd" authn
  options set, using a "realm" name specified using the "realm" authn
  option.
* If the daemon is started with the "htpasswd" authn option, it will
  call out to an "htpasswd" helper binary to check passwords using a
  specified file.
* If the daemon is started with the "libsasl2" authn option enabled and
  built with support for it, it will call the SASL library to check
  passwords.  The library can, in turn, be configured to ask saslauthd
  to check passwords using whatever facilities it's configured to use.
* If the daemon is built with GSSAPI support, it will check the default
  keytab file, or one specified using the "keytab" authn option, for
  acceptor creds, and if it finds them, will offer Negotiate
  authentication.
* If the daemon is started with the "local-auth" authn option enabled,
  it will attempt to authenticate clients by asking the kernel to
  identify the client user.  On Linux systems, this is done by reading
  the client's UnixCredentials, and is only supported for "unix"
  connections.
* If the daemon is started with the "certmap" authn option, it will
  attempt to use specified plugins to identify HTTPS clients by their
  client certificates.
* If the daemon is started with the "plugins" authn option, it will also
  call out to authenticators implemented by the specified Docker
  plugins.

* When a daemon responds to a client request with a 401 error code, the
  client will attempt to authenticate using whatever scheme the daemon
  offers as part of the error message and try again.  The client has
  built-in support for Basic, Negotiate, and Bearer schemes.
* When attempting Basic authentication, the client will use a user name
  passed in as the values for the "basic.username" authn option.  If one
  is not specified on the command line, it will attempt to read the
  value to use from the $DOCKER_AUTHN_USERID variable in the
  environment.  It will also attempt to read a password from the
  $DOCKER_AUTHN_PASSWORD variable in the environment.  If it's not able
  to determine both a user name and a password, and the client has a
  controlling terminal, it will then prompt for the values, unless the
  "interactive" authn option has been set to "false".
* When attempting Bearer authentication, the client will use a value
  passed in as the value for the "bearer.token" authn option.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Call into gssapi to check Negotiate auth tokens

Use the GSSAPI libraries to accept GSSAPI tokens, and offer the option
of using it as the Negotiate mechanism if we can find server credentials
to use at startup-time.

If a "keytab" authentication option is passed as a command-line option,
we'll attempt to register that as the Kerberos keytab to be used, also
at startup-time.

We don't set a service name, so we'll accept authentication as any
service for which Kerberos keys can be found in the keytab, but the
expectation is that clients will authenticate to the service named
"HTTP" on the host, since that's what the Negotiate spec says they
should use.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Pass along Bearer auth data

When a server offers Bearer authentication, check for a token specified
as an authentication option on the command-line, or in the environment
(preferring values from the command-line), and if we find one, use it.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind commented:
Rebased to take into account changes from #18472 (Basic and Bearer creds are now provided by having the DockerCli object set callbacks in its Client object) and #15365 (we pass information about authenticated users into the authorization logic, reorder the middleware initialization to ensure that authorization is called after authentication, and add a test that exercises both authentication and authorization).

nalind committed:
Clean up tagging of requirements for authn tests

Clean up the tagging of authentication integration tests, supplementing
checking if we have a dynamically-linked ELF binary with checking if we
have unresolved references to symbols exposed by GSSAPI and libsasl2, so
that we can run them properly in the CI.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

diogomonica commented on api/client/authn.go:
```diff
@@ -0,0 +1,108 @@
+	return "", fmt.Errorf("invalid authentication option %s", option)
+}
+
+func (cli *DockerCli) getBasic(challenge string, defusername string) (string, string, error) {
```
Should be `challenge, defusername string`. Also `defusername` doesn't tell me much as a variable name.

nalind replied:
Trying to pass in a default user name no longer made sense after the last rebase, so I'll replace this with just the realm name.  Thanks!

diogomonica commented on api/client/authn.go:
```diff
@@ -0,0 +1,108 @@
+// authentication option.
+func ValidateAuthnOpt(option string) (string, error) {
+	if strings.HasPrefix(option, "basic.username=") ||
+		strings.HasPrefix(option, "basic.password=") ||
```
I had this same concern in the design: I don't think we should allow passwords to be set as cmd-line arguments (I understand that `docker login` has the same problem, but we should definitely not propagate bad design decisions).

nalind replied:
I'll add a note to the docs about its use not being recommended.  We use it in the integration tests, though.

diogomonica replied:
I think we could as easily use env variables in the integration tests. I'm very much against allowing this behavior in the first place.

nalind replied:
Okay, removing the basic.password authentication option.

diogomonica commented on api/client/authn.go:
```diff
@@ -0,0 +1,108 @@
+	if username == "" {
+		username = defusername
+	}
+	defusername = username
```
Shouldn't we do a final check to make sure `defusername` is also empty and fail if so?

diogomonica commented on api/client/authn.go:
```diff
@@ -0,0 +1,108 @@
+			return string(line)
+		}
+
+		if defusername != "" {
```
Aren't `defusername` and `username` the same at this point? (line 36)

This is confusing because it seems that we are testing a conditional on the unmodified argument, when in fact it always gets overwritten earlier in the code.

diogomonica commented on api/client/authn.go:
```diff
@@ -0,0 +1,108 @@
+		username = readInput(cli.in, cli.out)
+		username = strings.Trim(username, " ")
+		if username == "" {
+			username = defusername
```
Is this a best effort attempt at taking the username from the cmd-line of env variable in the case the reason we are in this `if` is due to `password == ""` ?

nalind replied:
Yep.  The idea was to take a default from the client library, replacing it with a supplied value if one was supplied at the command line or via the environment.  If no password was similarly supplied, the client would only need to confirm the user name value.

diogomonica commented on api/client/authn.go:
```diff
@@ -0,0 +1,108 @@
+	return username, password, nil
+}
+
+func (cli *DockerCli) getBearer(challenge string) (string, error) {
```
This method takes a `challenge string` and doesn't use it for anything. Not sure at this point if we are trying to meet an interface somewhere else.

nalind replied:
Oh, right.  It probably makes more sense to pull out the realm from the Basic challenge, pass that in, and include it in the prompt if we have a value.

diogomonica commented on api/client/lib/authn.go:
```diff
@@ -0,0 +1,211 @@
+	// AuthCompleted, given a (possibly empty) WWW-Authenticate header from
+	// a successful response, should decide if the server's reply should be
+	// accepted.
+	authCompleted(cli *Client, challenge string, req *http.Request) (bool, error)
```
Shouldn't `authCompleted` operate on an `http.Response`?

nalind replied:
Hmm, it does look that way.  (They tend to only care about WWW-Authenticate, but that could change in the future.)  There's a shared code path in the negotiate client that expects to be able to read the server's host name out of the request, but that's easily fixed.  I'll change it.

diogomonica commented on api/client/lib/authn.go:
```diff
@@ -0,0 +1,211 @@
+
+// registerAuthResponder registers a function which will be called at startup
+// to create an authResponder.
+func registerAuthResponder(arc authResponderCreator) {
```
Not sure if this should be thread safe or not (access to `authReponderCreators`). It's probably not an issue.

nalind replied:
It should be - there wouldn't be any functional difference if it was covered by an RWMutex that we kept locked for read-only access.

diogomonica commented on api/client/lib/client.go:
```diff
@@ -30,13 +31,19 @@ type Client struct {
+	// a cookie jar, if we're meant to be using cookies
+	jar *cookiejar.Jar
+	// authResponders provide ways to authenticate to servers
+	authResponders *map[string]*[]authResponder
```
We should probably maintain consistency with the current parameters and do a `map[string]*[]authReponder`.

diogomonica commented on api/client/lib/authn.go:
```diff
@@ -0,0 +1,211 @@
+
+// Run through all of the registered responder creators and build a map of
+// lists of responders.
+func createAuthResponders() *map[string]*[]authResponder {
```
No entirely convinced of the need the need for this to be a `*[]authReponder`.

diogomonica commented on api/client/lib/authn.go:
```diff
@@ -0,0 +1,211 @@
+// authResponderCreator either creates a new authResponder, or returns nil
+type authResponderCreator func() authResponder
+
+var authResponderCreators = []authResponderCreator{}
```
What is the need for a `authResponderCreator`? It seems like `authResponder` always receives the parameters it needs to provide a valid authentication response, and the `creator` is a shim on top so we can iterate over `authResponder` and create a division by scheme. Why aren't authResponder's registered directly in a map that splits them by scheme, and we can kill `authResponderCreator` altogether.

nalind replied:
If we switch from storing arrays in the map to single objects, the main purpose of using functions to create objects rather than using singletons is to let authentication implementations keep their state in the Client object rather than making it global - I'm thinking mainly of caching the creds for Basic, and partially-established contexts in Negotiate.

diogomonica commented on api/client/lib/authn.go:
```diff
@@ -0,0 +1,211 @@
+			scheme := strings.ToLower(responder.scheme())
+			if ars[scheme] != nil {
+				slice := append(*(ars[scheme]), responder)
+				ars[scheme] = &slice
```
Yeah, not being a pointer would have turned this into a prettier one-liner.

diogomonica commented on api/client/lib/authn.go:
```diff
@@ -0,0 +1,211 @@
+// it, but we'd still have to do the work for httputil.ClientConn, so we don't
+// set it for http.Clients to avoid duplicating cookies in requests.
+func (cli *Client) doWithCookies(doer doer, req *http.Request) (*http.Response, error) {
+	if localhost, ok := os.Hostname(); ok == nil && cli.proto == "unix" {
```
this should be `err` not ok.

nalind replied:
Doh.  Fixing.

diogomonica commented on api/client/lib/authn.go:
```diff
@@ -0,0 +1,211 @@
+
+// doWithAuthn calls Do(), and handles any "unauthorized" errors
+// which it returns by retrying it with authentication.
+func (cli *Client) doWithAuthn(doer doer, req *http.Request, body []byte) (*http.Response, error) {
```
I don't think we should have methods with 7 indentation levels and 3 nested for's. This is really hard to read and should be refactored somehow.

nalind replied:
I think that between cutting things back to one client implementation per scheme and a little bit of restructuring, that can be improved.

nalind replied:
Okay, I think it looks better now than before.  Refactoring was necessary on the client side when the api client lib changes landed anyway.

runcom committed:
Use hack/vendor.sh to import deps

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

nalind committed:
Prefer using dynamic binaries in integration tests

Some of the newer integration tests depend on features that we don't
build into the static binary.  We overlooked this possibility when we
switched the top-level Makefile over to preferring dynbinary over
binary, but that wasn't enough for the CI builds.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Vendor golang.org/x/sys

We try to read the remote user's identity using the socket on Unix
systems, but the syscall package doesn't always include the functions
that we use to do so (specifically, when we build with gccgo).

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind committed:
Use golang.org/x/sys/unix instead of syscall

Use the "golang.org/x/sys/unix" package to read a user's Unix
credentials, instead of the "syscall" package, which might not provide
the functions for doing so.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

shykes commented:
Thanks @nalind. I'm concerned about the impact on performance. Could you perhaps run some benchmarks against this with a few example auth mechanisms, and share the results?

Also, it feelslike the documentation is lacking. Could you share the final state documentationfor this.

Thanks again for this super cool contribution.

endophage commented:
Requests because this is so large as to be almost impossible to review:

- It seems like a lot of files have a single line change that adds `cli.InstallCommonFlags(cmd)`. That's creating a lot of noise and preventing github from displaying the full PR. Could those changes be moved into a separate PR?
- It feels like the plugin code should be PR'd separately and reviewed prior to any specific implementations (it's useful to have a concrete implementation to review the plugin against so maybe we choose basic_htpasswd as the reference plugin implementation to be included with the main plugin PR).
- If the prior bullet makes sense, it would then be great to break out additional plugin implementations into one PR each so they can be evaluated on their own merits.

nalind commented:
@endophage The changes under api/client are copied over from an earlier version of https://github.com/docker/engine-api/pull/10, which contains only the client side bits.  They're still in here only so that we can build and run tests -- once it's merged there and the vendored copy is updated, they'll fall out of this PR.

@shykes I think docs/reference/commandline/daemon.md could stand to describe the options that are recognized, but beyond that I can't tell what else is missing.  But then I've been staring at this for a while.  Are you point to particular sections or areas that you're thinking still need fleshing out?

@endophage I'd be as happy to drop plugins as anything, since they open up the possibility that we'd need to come up with a way to let plugins bypass authentication if they ever tried to call back into the daemon (for now, the answer for that probably looks like "don't use remote plugins, and turn on local-auth").  Most of the mechanisms aren't implemented as plugins, but we exercise the plugin interface with a very basic "Bearer" plugin during integration testing to keep things sane.

nalind commented:
@endophage Whoops, on another look, that's not bits that would be in engine-api after all.

nalind commented:
Opened https://github.com/docker/docker/pull/19993 with the InstallCommonFlags() changes, renamed to AddCommonFlags() because it's doing a job that's similar to addTrustedFlags()'s.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "daemonauth" git@github.com:nalind/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~26
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/838/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.3 m	docker.exe
  0.0%
  3.6%
  7.3%
 10.9%
 14.6%
 18.2%
 21.9%
 25.5%
 29.2%
 32.8%
 36.5%
 40.1%
 43.8%
 47.4%
 51.1%
 54.7%
 58.4%
 62.0%
 65.7%
 69.3%
 73.0%
 76.6%
 80.3%
 83.9%
 87.6%
 91.2%
 94.9%
 98.5%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.78 m   54.78 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1148941920 Bytes/sec.
   Speed :            65742.984 MegaBytes/min.
   Ended : Wednesday, February 10, 2016 5:28:02 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-da
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1751/console):

~~~console

---
ndles/1.11.0-dev/test-integration-cli)
---> Making bundle: .ensure-syscall-test (in bundles/1.11.0-dev/test-integration-cli)
+ go test -test.timeout=180m -check.v github.com/docker/docker/integration-cli
# testmain
/usr/bin/ld: $WORK/github.com/mattn/libgo-sqlite3.a(sqlite3-binding.o): undefined reference to symbol 'pthread_mutexattr_settype@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
FAIL	github.com/docker/docker/integration-cli [build failed]
---> Making bundle: .integration-daemon-stop (in bundles/1.11.0-dev/test-integration-cli)
+++++ cat bundles/1.11.0-dev/test-integration-cli/docker.pid
++++ kill 5342
++++ /etc/init.d/apparmor stop
Clearing AppArmor profiles cache:.
All profile caches have been cleared, but no profiles have been unloaded.
Unloading profiles will leave already running processes permanently
unconfined, which can lead to unexpected situations.

To set a process to complain mode, use the command line tool

---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/144/console):

~~~console

---
engine-api/client/hijack.go
 M vendor/src/github.com/docker/engine-api/client/interface.go
 D vendor/src/github.com/docker/engine-api/client/middleware.go
 M vendor/src/github.com/docker/engine-api/client/request.go
 M vendor/src/golang.org/x/sys/unix/syscall_linux.go
 M vendor/src/golang.org/x/sys/unix/syscall_solaris.go
 M vendor/src/golang.org/x/sys/unix/zsyscall_solaris_amd64.go

Please vendor your package with ./hack/vendor.sh.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson8593997546256872897.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/857/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.3 m	docker.exe
  0.0%
  3.6%
  7.3%
 10.9%
 14.6%
 18.2%
 21.9%
 25.5%
 29.2%
 32.8%
 36.5%
 40.1%
 43.8%
 47.4%
 51.1%
 54.7%
 58.4%
 62.0%
 65.7%
 69.3%
 73.0%
 76.6%
 80.3%
 83.9%
 87.6%
 91.2%
 94.9%
 98.5%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.78 m   54.78 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           941755672 Bytes/sec.
   Speed :           53887.691 MegaBytes/min.
   Ended : Wednesday, February 10, 2016 6:55:09 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-cb80
---

~~~

nalind commented:
Per discussion in PR#19993, moving the addition of the --authn-opt flag to the common flags, and adjusting the docs and tests to pass the flag to clients correctly (i.e., before the client command).

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/917/console):

~~~console

---
111e45.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Thu Feb 11 15:44:56 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Thu Feb 11 15:44:57 CUT 2016 (0m 54s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson8947572308710169798.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/165/console):

~~~console

---
engine-api/client/hijack.go
 M vendor/src/github.com/docker/engine-api/client/interface.go
 D vendor/src/github.com/docker/engine-api/client/middleware.go
 M vendor/src/github.com/docker/engine-api/client/request.go
 M vendor/src/golang.org/x/sys/unix/syscall_linux.go
 M vendor/src/golang.org/x/sys/unix/syscall_solaris.go
 M vendor/src/golang.org/x/sys/unix/zsyscall_solaris_amd64.go

Please vendor your package with ./hack/vendor.sh.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson7860011402397900634.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/923/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.3 m	docker.exe
  0.0%
  3.6%
  7.2%
 10.9%
 14.5%
 18.2%
 21.8%
 25.5%
 29.1%
 32.8%
 36.4%
 40.1%
 43.7%
 47.4%
 51.0%
 54.7%
 58.3%
 62.0%
 65.6%
 69.3%
 72.9%
 76.6%
 80.2%
 83.9%
 87.5%
 91.2%
 94.8%
 98.5%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.79 m   54.79 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1249139304 Bytes/sec.
   Speed :            71476.324 MegaBytes/min.
   Ended : Thursday, February 11, 2016 5:00:54 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-99d
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/175/console):

~~~console

---
engine-api/client/hijack.go
 M vendor/src/github.com/docker/engine-api/client/interface.go
 D vendor/src/github.com/docker/engine-api/client/middleware.go
 M vendor/src/github.com/docker/engine-api/client/request.go
 M vendor/src/golang.org/x/sys/unix/syscall_linux.go
 M vendor/src/golang.org/x/sys/unix/syscall_solaris.go
 M vendor/src/golang.org/x/sys/unix/zsyscall_solaris_amd64.go

Please vendor your package with ./hack/vendor.sh.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson5279696206773317081.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/979/console):

~~~console

---
2T16:03:34Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End of cleanup
INFO: Ended at Fri Feb 12 16:03:34 CUT 2016 (1m 15s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson5234087706729695417.sh
+ set +e
+ set +x
INFO: Nuking /d/CI
time="2016-02-12T16:03:40Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End of clean
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/179/console):

~~~console

---
engine-api/client/hijack.go
 M vendor/src/github.com/docker/engine-api/client/interface.go
 D vendor/src/github.com/docker/engine-api/client/middleware.go
 M vendor/src/github.com/docker/engine-api/client/request.go
 M vendor/src/golang.org/x/sys/unix/syscall_linux.go
 M vendor/src/golang.org/x/sys/unix/syscall_solaris.go
 M vendor/src/golang.org/x/sys/unix/zsyscall_solaris_amd64.go

Please vendor your package with ./hack/vendor.sh.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson305662342800336585.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/187/console):

~~~console

---

 D vendor/src/github.com/docker/engine-api/types/strslice/strslice_test.go
 D vendor/src/github.com/docker/engine-api/types/time/timestamp_test.go
 D vendor/src/github.com/docker/engine-api/types/versions/README.md
 M vendor/src/golang.org/x/sys/unix/syscall_linux.go
 M vendor/src/golang.org/x/sys/unix/syscall_solaris.go
 M vendor/src/golang.org/x/sys/unix/zsyscall_solaris_amd64.go

Please vendor your package with ./hack/vendor.sh.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson8886861730920530886.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15000/console):

~~~console

---
.TestCreateWhenCertExpired (Currently changes system time, causing instability)
SKIP: docker_cli_pull_trusted_test.go:65: DockerTrustSuite.TestPullWhenCertExpired (Currently changes system time, causing instability)
SKIP: docker_cli_run_test.go:3212: DockerTrustSuite.TestRunWhenCertExpired (Currently changes system time, causing instability)
PASS: docker_cli_build_test.go:5664: DockerTrustSuite.TestTrustedBuild	1.382s

----------------------------------------------------------------------
FAIL: docker_cli_build_test.go:5695: DockerTrustSuite.TestTrustedBuildUntrustedTag

docker_cli_build_test.go:5712:
    c.Fatalf("Unexpected output on trusted build with untrusted tag:\n%s", out)
... Error: Unexpected output on trusted build with untrusted tag:
time="2016-02-16T05:01:17Z" level=debug msg="reading certificate directory: /go/src/github.com/docker/docker/bundles/1.11.0-dev/test-integration-cli/fake-HOME/.docker/tls/localhost:4443" 
time="2016-02-16T05:01:17Z" level=debug msg="Failed to initialize PKCS11 environment: no library found." 
time="2016-02-16T05:0
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23786/console):

~~~console

---
.TestCreateWhenCertExpired (Currently changes system time, causing instability)
SKIP: docker_cli_pull_trusted_test.go:65: DockerTrustSuite.TestPullWhenCertExpired (Currently changes system time, causing instability)
SKIP: docker_cli_run_test.go:3212: DockerTrustSuite.TestRunWhenCertExpired (Currently changes system time, causing instability)
PASS: docker_cli_build_test.go:5664: DockerTrustSuite.TestTrustedBuild	1.512s

----------------------------------------------------------------------
FAIL: docker_cli_build_test.go:5695: DockerTrustSuite.TestTrustedBuildUntrustedTag

docker_cli_build_test.go:5712:
    c.Fatalf("Unexpected output on trusted build with untrusted tag:\n%s", out)
... Error: Unexpected output on trusted build with untrusted tag:
time="2016-02-16T05:06:50Z" level=debug msg="reading certificate directory: /go/src/github.com/docker/docker/bundles/1.11.0-dev/test-integration-cli/fake-HOME/.docker/tls/localhost:4443" 
time="2016-02-16T05:06:50Z" level=debug msg="Adding cert with certID: db52734ceed96a72b7744a438d032c86c8ecde27b89460811050ce545e
---

~~~

calavera commented:
Dear @docker/security,

Some of the people in @docker/core-engine-maintainers met last week to discuss this contribution and some others. We didn't find a very compelling reason to move this forward, authentication can easily be done outside the docker engine as a proxy.

We'd like to know if you feel very strong about making this part of the Docker Engine or not. We'd also like to know if anyone would like to take this change the extra mile needed to make it production ready. Otherwise, we'll consider this a `won't fix` and close the pull request without merging it.

:green_heart:

rhatdan commented:
> Some of the people in @docker/core-engine-maintainers met last week to discuss this contribution and > some others. We didn't find a very compelling reason to move this forward, authentication can easily be > done outside the docker engine as a proxy.

How?  We need to know who is directing the action on the engine so that RBAC will work.  How can I state that dwalsh is allowed to only start/stop container X,Y,Z and now allowed to create a new container.  Or not allowed to create a container using --privileged?

We need proper auditing/logging in the docker daemon.  Do you want to have a full tool standing in front of the docker socket and listening on the docker port that will look and understand all incoming connections to the docker ports and log all this information outside of the docker service?  Seems like a lot of work and likely to be out of sync with the docker daemon engine.

> We'd like to know if you feel very strong about making this part of the Docker Engine or not. We'd also > like to know if anyone would like to take this change the extra mile needed to make it production ready. > Otherwise, we'll consider this a won't fix and close the pull request without merging it.

Having an incredibly powerful daemon that basically gives any process that can connect to it full root on a system, without any logging/auditing of who connected is a security disaster, and will force us to always recommend to users to never open access to these sockets to anything that is not running as root.  It will also hinder the use of the docker daemon in government institutions that require real logging/auditing.  

You also eliminate any chance to allow meaningful RBAC of the docker daemon.  Having authorization without authentication ...

>We'd like to know if you feel very strong about making this part of the Docker Engine or not. We'd also 
> like to know if anyone would like to take this change the extra mile needed to make it production ready. > Otherwise, we'll consider this a won't fix and close the pull request without merging it.

What do you mean the extra mile?  Can we get feedback on what you don't like about the patches.  I am sure the people working on this would welcome feedback.

cpuguy83 commented:
I agree we should have authentication, but with anything more than basic auth + PAM should be done in plugins.

nalind commented:
We can do some of what we're currently doing natively in plugins, though of course it makes deploying any of this more of a hassle on systems without socket activation.
One tricky part is that I'm not 100% sure we can pass descriptors to docker plugins, which would be required for local-auth to be doable as a plugin.
There are certain things that you want to check first (local-auth if it's possible, then certificate-based identity if it's enabled, since neither requires the client to learn to do anything it can't already), and turning every possibility into a plugin makes getting the ordering right the admin's problem, which is not really desirable.

cpuguy83 commented:
@nalind I'm not sure how I feel about the TLS cert inspection yet... but the unix sock one feels like it's not a good solution and generally a hack (albeit a glorious one!)... the rampant use of`reflect` being a pretty good clue on that, the result of which is largely unmaintainable code which also doesn't work for the poor souls using gccgo.

I would like to make sure the sec team is part of the discussion before we move anything forward here.
Beyond that I would like to see much of this distilled down into a simple auth interface that we can easily wire up various backends to via plugins, and then decide what (if any) we should include as a built-in.

adrianp-sti commented:
I'm a huge fan of this work and personally would love to see it merged.  @rhatdan pretty much hit the nail on the head with the reasons why.  To my mind the authZ functionality (at least in terms of granularity) is severely limited without authN - I'm specifically talking a proper RBAC implementation here.  I don't understand the logic of including authZ framework but then say go build yourself some authN.

In regards to the form it takes if merged, I like that the included modules (i.e. htpasswd, SASL etc.)  give users a low barrier to entry in that there is no Go coding required and you can get authN up and running. I think this is important if we want to encourage people to use it. If given a choice I'd take the way the patch is at the moment with the supplied modules and the ability to write your own plugin.  At a bare minimum level it'd be the ability to write your own plugin.

jfrazelle commented:
IMO merging this patch as is would be a grave mistake. It is not a question
in my mind wrt _if_ we want auth, the patch in its current form would be a
huge burden to maintainers. It is quite frankly a mess. I would not expect
someone to merge a patch like this into any project.

On Sunday, February 28, 2016, AdrianP <notifications@github.com> wrote:

> I'm a huge fan of this work and personally would love to see it merged.
> @rhatdan <https://github.com/rhatdan> pretty much hit the nail on the
> head with the reasons why. To my mind the authZ functionality (at least in
> terms of granularity) is severely limited without authN - I'm specifically
> talking a proper RBAC implementation here. I don't understand the logic of
> including authZ framework but then say go build yourself some authN.
>
> In regards to the form it takes if merged, I like that the included
> modules (i.e. htpasswd, SASL etc.) give users a low barrier to entry in
> that there is no Go coding required and you can get authN up and running. I
> think this is important if we want to encourage people to use it. If given
> a choice I'd take the way the patch is at the moment with the supplied
> modules and the ability to write your own plugin. At a bare minimum level
> it'd be the ability to write your own plugin.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/18514#issuecomment-189919668>.
>


-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

jfrazelle commented:
We can sit here and ping @docker/security until the cows come home but they
will not respond. They also aren't the ones that will have the burden on
maintaining this 9K lines of code.
I'm closing seeing as no maintainer has defended the patch after
https://github.com/docker/docker/pull/18514#issuecomment-187942565

> We'd also like to know if anyone would like to take this change the extra
mile needed to make it production ready.

But I'd still urge someone to step up if they believe in the above.


On Sunday, February 28, 2016, Jessica Frazelle <me@jessfraz.com> wrote:

> IMO merging this patch as is would be a grave mistake. It is not a
> question in my mind wrt _if_ we want auth, the patch in its current form
> would be a huge burden to maintainers. It is quite frankly a mess. I would
> not expect someone to merge a patch like this into any project.
>
> On Sunday, February 28, 2016, AdrianP <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> I'm a huge fan of this work and personally would love to see it merged.
>> @rhatdan <https://github.com/rhatdan> pretty much hit the nail on the
>> head with the reasons why. To my mind the authZ functionality (at least in
>> terms of granularity) is severely limited without authN - I'm specifically
>> talking a proper RBAC implementation here. I don't understand the logic of
>> including authZ framework but then say go build yourself some authN.
>>
>> In regards to the form it takes if merged, I like that the included
>> modules (i.e. htpasswd, SASL etc.) give users a low barrier to entry in
>> that there is no Go coding required and you can get authN up and running. I
>> think this is important if we want to encourage people to use it. If given
>> a choice I'd take the way the patch is at the moment with the supplied
>> modules and the ability to write your own plugin. At a bare minimum level
>> it'd be the ability to write your own plugin.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/docker/docker/pull/18514#issuecomment-189919668>.
>>
>
>
> --
>
>
> Jessie Frazelle
> 4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
> pgp.mit.edu
> <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>
>
>

-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

NathanMcCauley commented:
@nalind is working on a slimmed down version of this PR that will be a better starting point for adding authentication.

It'll start with something like cert-based authentication mappers (since we already support cert authentication) and basic auth.

thaJeztah commented:
Thanks for the heads-up @NathanMcCauley

thaJeztah commented:
oh and look who just came home!

![cow-came-home-well-sort-of](https://cloud.githubusercontent.com/assets/1804568/13404948/7954a450-df1c-11e5-96b2-77206cfcd4e2.jpg)
